Source/Core/DolphinQt/FIFO/FIFOAnalyzer.cpp

```diff
@@ -164,24 +173,124 @@ void FIFOAnalyzer::UpdateTree()
   }
 }
 
-static std::string GetPrimitiveName(u8 cmd)
+namespace
+{
+class DetailCallback : public OpcodeDecoder::Callback
 {
-  if ((cmd & 0xC0) != 0x80)
+public:
+  DetailCallback(CPState cpmem) : m_cpmem(cpmem) {}
+
+  void OnCP(u8 command, u32 value) override
   {
-    PanicAlertFmt("Not a primitive command: {:#04x}", cmd);
-    return "";
+    const auto [name, desc] = GetCPRegInfo(command, value);
+    ASSERT(!name.empty());
+
+    text = QStringLiteral("CP  %1  %2  %3")
+               .arg(command, 2, 16, QLatin1Char('0'))
+               .arg(value, 8, 16, QLatin1Char('0'))
+               .arg(QString::fromStdString(name));
   }
-  const u8 vat = cmd & OpcodeDecoder::GX_VAT_MASK;  // Vertex loader index (0 - 7)
-  const u8 primitive =
-      (cmd & OpcodeDecoder::GX_PRIMITIVE_MASK) >> OpcodeDecoder::GX_PRIMITIVE_SHIFT;
-  static constexpr std::array<const char*, 8> names = {
-      "GX_DRAW_QUADS",        "GX_DRAW_QUADS_2 (nonstandard)",
-      "GX_DRAW_TRIANGLES",    "GX_DRAW_TRIANGLE_STRIP",
-      "GX_DRAW_TRIANGLE_FAN", "GX_DRAW_LINES",
-      "GX_DRAW_LINE_STRIP",   "GX_DRAW_POINTS",
-  };
-  return fmt::format("{} VAT {}", names[primitive], vat);
-}
+
+  void OnXF(u16 address, u8 count, const u8* data) override
+  {
+    const auto [name, desc] = GetXFTransferInfo(address, count, data);
+    ASSERT(!name.empty());
+
+    const u32 command = address | (count << 16);
+
+    text = QStringLiteral("XF  %1  ").arg(command, 8, 16, QLatin1Char('0'));
+
+    for (u8 i = 0; i < count; i++)
+    {
+      const u32 value = Common::swap32(&data[i * 4]);
+
+      text += QStringLiteral("%1 ").arg(value, 8, 16, QLatin1Char('0'));
+    }
+
+    text += QStringLiteral("  ") + QString::fromStdString(name);
+  }
+
+  void OnBP(u8 command, u32 value) override
+  {
+    const auto [name, desc] = GetBPRegInfo(command, value);
+    ASSERT(!name.empty());
+
+    text = QStringLiteral("BP  %1  %2  %3")
+               .arg(command, 2, 16, QLatin1Char('0'))
+               .arg(value, 6, 16, QLatin1Char('0'))
+               .arg(QString::fromStdString(name));
+  }
+  void OnIndexedLoad(u8 array, u32 index, u16 address, u8 size) override
+  {
+    const auto [desc, written] = GetXFIndexedLoadInfo(array, index, address, size);
+    text = QStringLiteral("LOAD INDX %1   %2")
+               .arg("ABCD"[array - ARRAY_XF_A])
+               .arg(QString::fromStdString(desc));
+  }
+  void OnPrimitiveCommand(OpcodeDecoder::Primitive primitive, u8 vat, u32 vertex_size,
+                          u16 num_vertices, const u8* vertex_data) override
+  {
+    const auto name = fmt::to_string(primitive);
```

Comment:
```suggestion
    const auto name = QString::number(primitive);
```

Avoids the need for conversion from std::string to QString.

Reply:
`fmt::to_string` is being used to go through the EnumFormatter for e.g. `GX_DRAW_QUADS`, not to create a number.  So that wouldn't do the right thing.